### PR TITLE
Discover new plugins with a more efficient technique

### PR DESCRIPTION
**Pull request by MohamedAli1995**
issue: Optimize plugins discovery #1733
brief about the proposed technique :
1- Load all plugins once.
2-Every plugin type should register its own plugins that are already loaded.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -156,7 +156,6 @@ set (CMAKE_BUILD_WITH_INSTALL_RPATH FALSE)
 #--------------------------------------------------------------------------------------------------
 # Platform-specific settings.
 #--------------------------------------------------------------------------------------------------
-
```
Comment (dictoon) [inline CMakeLists.txt]: Revert this change.

File: src/appleseed/renderer/modeling/entity/entityfactoryregistrar.cpp [hunk 2]
```diff
@@ -72,6 +72,50 @@ EntityFactoryRegistrar::~EntityFactoryRegistrar()
...
+            search_path = search_paths.get_root_path().c_str() / search_path;
+
+        // Only consider directories.
+        if (!bf::exists(search_path) || !bf::is_directory(search_path))
+            continue;
+
+        // Iterate over all files in this directory.
+        for (bf::directory_iterator j(search_path), f; j != f; ++j)
+        {
+
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/entity/entityfactoryregistrar.cpp]: Remove blank line.

File: src/appleseed/renderer/modeling/entity/entityfactoryregistrar.cpp [hunk 3]
```diff
@@ -72,6 +72,50 @@ EntityFactoryRegistrar::~EntityFactoryRegistrar()
...
+            // Only consider shared library files.
+            if (!bf::is_regular_file(*j) ||
+                j->path().extension() != foundation::SharedLibrary::get_default_file_extension())
+                continue;
+
+            const std::string plugin_path = j->path().string();
+            
+            //load plugin into memory
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/entity/entityfactoryregistrar.cpp]: Follow the commenting style:
- Insert a space after `//`
- Capitalize the first letter of the sentence.
- End the sentence with a point.

File: src/appleseed/renderer/modeling/entity/entityfactoryregistrar.cpp [hunk 4]
```diff
@@ -72,6 +72,50 @@ EntityFactoryRegistrar::~EntityFactoryRegistrar()
     delete impl;
 }
 
+void EntityFactoryRegistrar::discover_plugins(
+    const foundation::SearchPaths&          search_paths){
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/entity/entityfactoryregistrar.cpp]: Braces always go on their own line.

File: src/appleseed/renderer/modeling/entity/entityfactoryregistrar.h [hunk 5]
```diff
@@ -77,6 +88,9 @@ class APPLESEED_DLLSYMBOL EntityFactoryRegistrar
     struct Impl;
     Impl* impl;
 
+    static std::vector<std::pair<std::unique_ptr<foundation::SharedLibrary>, std::string> > loaded_libraries;
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/entity/entityfactoryregistrar.h]: Must not be static.

File: src/appleseed/renderer/modeling/entity/registerentityfactories.h [hunk 6]
```diff
@@ -56,76 +56,43 @@ namespace renderer
...
 
     const std::string entry_point_name =
         foundation::format("appleseed_create_{0}_factory", EntityTraits<Entity>::get_entity_type_name());
 
-    // Iterate over all search paths.
-    for (size_t i = 0, e = search_paths.get_path_count(); i < e; ++i)
-    {
-        bf::path search_path(search_paths.get_path(i));
+    const std::string entity_type_name =
+            foundation::lower_case(EntityTraits<Entity>::get_human_readable_entity_type_name());
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/entity/registerentityfactories.h]: Fix indentation.

File: src/appleseed/renderer/modeling/entity/registerentityfactories.h [hunk 7]
```diff
@@ -56,76 +56,43 @@ namespace renderer
...
 
-        // Make the search path absolute if it isn't already.
-        if (!search_path.is_absolute() && search_paths.has_root_path())
-            search_path = search_paths.get_root_path().c_str() / search_path;
+    for (int i = 0; i < loaded_libraries.size(); i++)
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/entity/registerentityfactories.h]: Type mismatch, this won't compile on Windows. Use `size_t`.

File: src/appleseed/renderer/modeling/project/project.cpp [hunk 8]
```diff
@@ -38,6 +38,8 @@
 #include "renderer/modeling/environment/environment.h"
 #include "renderer/modeling/environmentedf/environmentedf.h"
 #include "renderer/modeling/environmentshader/environmentshader.h"
+#include "renderer/modeling/entity/entityfactoryregistrar.h"
+
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/project/project.cpp]: Remove blank line.

File: src/appleseed/renderer/modeling/aov/aovfactoryregistrar.cpp [hunk 9]
```diff
@@ -62,19 +62,23 @@ struct AOVFactoryRegistrar::Impl
     Registrar<IAOVFactory> m_registrar;
 };
 
-AOVFactoryRegistrar::AOVFactoryRegistrar(const SearchPaths& search_paths)
-  : impl(new Impl())
+AOVFactoryRegistrar::AOVFactoryRegistrar()
+  : impl(new Impl()),
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/aov/aovfactoryregistrar.cpp]: Comma goes on the next line. Check how it's done in other classes.

File: src/appleseed/renderer/modeling/aov/aovfactoryregistrar.cpp [hunk 10]
```diff
@@ -62,19 +62,23 @@ struct AOVFactoryRegistrar::Impl
...
+    m_loaded_libraries(boost::make_shared<loaded_libs_container>())
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/aov/aovfactoryregistrar.cpp]: Let's use C++11's `std::make_shared<>()` and co. rather than Boost.

File: src/appleseed/renderer/modeling/entity/entityfactoryregistrar.h [hunk 11]
```diff
@@ -50,13 +59,21 @@ namespace renderer      { class Plugin; }
 namespace renderer
 {
 
+//define loaded libraries container structure
+typedef  std::vector<std::pair<std::unique_ptr<foundation::SharedLibrary>, std::string> >  loaded_libs_container;
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/entity/entityfactoryregistrar.h]: As per our [naming conventions](https://github.com/appleseedhq/appleseed/wiki/Coding-Philosophy-and-Guidelines#naming-conventions) (please read them) that should be `LoadedLibsContainer`.

Also note that in C++11, you no longer need to insert spaces between `>`.

Finally I think the comment doesn't add much and could be removed.

File: src/appleseed/renderer/modeling/aov/aovfactoryregistrar.cpp [hunk 12]
```diff
@@ -62,19 +62,23 @@ struct AOVFactoryRegistrar::Impl
...
 AOVFactoryRegistrar::~AOVFactoryRegistrar()
 {
     delete impl;
 }
 
-void AOVFactoryRegistrar::reinitialize(const SearchPaths& search_paths)
+void AOVFactoryRegistrar::reinitialize(
+    const boost::shared_ptr<loaded_libs_container> loaded_libraries)
 {
+    m_loaded_libraries = loaded_libraries;
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/aov/aovfactoryregistrar.cpp]: I don't get it: `reinitialize()` is called with `m_loaded_libraries` in argument, and this gets assigned again to `m_loaded_libraries`?

File: src/appleseed/renderer/modeling/aov/aovfactoryregistrar.h [hunk 13]
```diff
@@ -36,10 +36,21 @@
 #include "foundation/utility/api/apiarray.h"
 #include "foundation/utility/autoreleaseptr.h"
 #include "foundation/utility/searchpaths.h"
+#include "foundation/platform/sharedlibrary.h"
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/aov/aovfactoryregistrar.h]: Order alphabetically.

File: src/appleseed/renderer/modeling/aov/aovfactoryregistrar.h [hunk 14]
```diff
@@ -36,10 +36,21 @@
...
 
 // appleseed.main headers.
 #include "main/dllsymbol.h"
 
+// Boost headers.
+#include "boost/filesystem.hpp"
+#include "boost/make_shared.hpp"
+
+//standard headers
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/aov/aovfactoryregistrar.h]: Follow our conventions for comments: `// Standard headers.`

File: src/appleseed/renderer/modeling/aov/aovfactoryregistrar.h [hunk 15]
```diff
@@ -36,10 +36,21 @@
...
+#include <string>
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/aov/aovfactoryregistrar.h]: Order alphabetically.

File: src/appleseed/renderer/modeling/aov/aovfactoryregistrar.h [hunk 16]
```diff
@@ -65,14 +76,14 @@ class APPLESEED_DLLSYMBOL AOVFactoryRegistrar
     typedef AOVFactoryArray FactoryArrayType;
 
     // Constructor.
-    explicit AOVFactoryRegistrar(
-        const foundation::SearchPaths& search_paths = foundation::SearchPaths());
+    explicit AOVFactoryRegistrar();
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/aov/aovfactoryregistrar.h]: Remove `explicit` as it's no longer required. But note for the record that I wish that C++ had explicit by default! :)

File: src/appleseed/renderer/modeling/aov/aovfactoryregistrar.h [hunk 17]
```diff
@@ -84,6 +95,9 @@ class APPLESEED_DLLSYMBOL AOVFactoryRegistrar
     struct Impl;
     Impl* impl;
 
+    // Pointer to the loaded libraries in memory.
+    boost::shared_ptr<loaded_libs_container> m_loaded_libraries;
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/aov/aovfactoryregistrar.h]: I wonder if we need to store loaded libraries in each registrar.

File: src/appleseed/renderer/modeling/bssrdf/bssrdffactoryregistrar.h [hunk 18]
```diff
@@ -36,10 +36,16 @@
 #include "foundation/utility/api/apiarray.h"
 #include "foundation/utility/autoreleaseptr.h"
 #include "foundation/utility/searchpaths.h"
+#include "foundation/platform/sharedlibrary.h"
 
 // appleseed.main headers.
 #include "main/dllsymbol.h"
 
+//standard headers
+#include <string>
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/bssrdf/bssrdffactoryregistrar.h]: It doesn't look like `<string>` and `<vector>` are required, please remove them. Moreover you're including `<memory>` which makes sense but then you're using `boost::shared_ptr<>` which is completely unrelated (and for that you're missing the header). Switch to `std::shared_ptr<>`.

File: src/appleseed/renderer/modeling/entity/entityfactoryregistrar.cpp [hunk 19]
```diff
@@ -72,6 +69,54 @@ EntityFactoryRegistrar::~EntityFactoryRegistrar()
...
+                j->path().extension() != foundation::SharedLibrary::get_default_file_extension())
+                continue;
+
+            const std::string plugin_path = j->path().string();
+            
+            // Load plugin into memory. 
+            try
+            {
+                unique_ptr<foundation::SharedLibrary> library(new foundation::SharedLibrary(plugin_path.c_str()));
+                lb->push_back(std::make_pair(std::move(library),plugin_path));
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/entity/entityfactoryregistrar.cpp]: Insert space after `,`.

File: src/appleseed/renderer/modeling/entity/entityfactoryregistrar.h [hunk 20]
```diff
@@ -30,18 +30,27 @@
 #define APPLESEED_RENDERER_MODELING_ENTITY_ENTITYFACTORYREGISTRAR_H
 
 // appleseed.renderer headers.
+#include "renderer/global/globallogger.h"
 #include "renderer/modeling/entity/entitytraits.h"
 
 // appleseed.foundation headers.
 #include "foundation/core/concepts/noncopyable.h"
 #include "foundation/utility/autoreleaseptr.h"
+#include "foundation/platform/sharedlibrary.h"
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/entity/entityfactoryregistrar.h]: Order alphabetically.

File: src/appleseed/renderer/modeling/entity/registerentityfactories.h [hunk 21]
```diff
@@ -43,89 +43,60 @@
...
-    // Iterate over all search paths.
-    for (size_t i = 0, e = search_paths.get_path_count(); i < e; ++i)
-    {
-        bf::path search_path(search_paths.get_path(i));
-
-        // Make the search path absolute if it isn't already.
-        if (!search_path.is_absolute() && search_paths.has_root_path())
-            search_path = search_paths.get_root_path().c_str() / search_path;
+    const std::string entity_type_name =
+            foundation::lower_case(EntityTraits<Entity>::get_human_readable_entity_type_name());
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/entity/registerentityfactories.h]: Fix indentation.

File: src/appleseed/renderer/modeling/aov/aovfactoryregistrar.h (hunk 17 above)
Comment (MohamedAli1995) [reply src/appleseed/renderer/modeling/aov/aovfactoryregistrar.h]: I don't think we need, but I thought it will be good for a future work, although reinitialize is called in the constructor of each plugin type which doesn't make sense after the changes I have made, 
so probably until now we needn't to store loaded libraries

File: src/appleseed/renderer/modeling/bssrdf/bssrdffactoryregistrar.h (hunk 18 above)
Comment (MohamedAli1995) [reply src/appleseed/renderer/modeling/bssrdf/bssrdffactoryregistrar.h]: I was searching for the difference between boost::shared_ptr<> and std::shared_ptr<> , thanks I will!

**Comment by dictoon** [comment]
Abandonned by the author.
